spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/client/OAuth2ClientAutoConfiguration.java

```diff
@@ -0,0 +1,182 @@
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication.Type;
+import org.springframework.boot.autoconfigure.condition.SpringBootCondition;
+import org.springframework.boot.autoconfigure.security.SecurityAutoConfiguration;
+import org.springframework.boot.context.properties.bind.Bindable;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.ConditionContext;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.ConfigurationCondition;
+import org.springframework.core.annotation.Order;
+import org.springframework.core.env.Environment;
+import org.springframework.core.type.AnnotatedTypeMetadata;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfiguration;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.config.annotation.web.configurers.oauth2.client.OAuth2LoginConfigurer;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationProperties;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Spring Security OAuth 2.0 / OpenID Connect 1.0 client support.
+ *
+ * @author Joe Grandja
+ * @since 2.0.0
+ */
+@Configuration
+@ConditionalOnWebApplication(type = Type.SERVLET)
+@ConditionalOnClass({EnableWebSecurity.class, ClientRegistration.class})
+@AutoConfigureBefore(SecurityAutoConfiguration.class)
+public class OAuth2ClientAutoConfiguration {
+
+	private static final String CLIENT_PROPERTY_PREFIX = "security.oauth2.client";
+
+	private static final String CLIENT_ID_PROPERTY = "client-id";
+
+	private static final String USER_INFO_URI_PROPERTY = "user-info-uri";
+
+	private static final String USER_NAME_ATTR_NAME_PROPERTY = "user-name-attribute-name";
+
+	static Set<String> getClientKeys(Environment environment) {
+		return getClientPropertiesByClient(environment).keySet();
+	}
+
+	static Map<String, Map> getClientPropertiesByClient(Environment environment) {
```

Comment:
It's unusual for an auto-configuration class to use the `Environment` directly. Ideally, this should be replaced with `@ConfigurationProperties` classes that I proposed above.

Reply:
See [comment](https://github.com/spring-projects/spring-boot/pull/10235#discussion_r138076857)